gobblin-core/src/main/java/gobblin/source/extractor/hadoop/OldApiHadoopFileInputSource.java

```diff
@@ -0,0 +1,214 @@
+      throw new IOException("No serialized FileSplit found in WorkUnitState " + workUnitState.getId());
+    }
+
+    JobConf jobConf = new JobConf(new Configuration());
+    for (String key : workUnitState.getPropertyNames()) {
+      jobConf.set(key, workUnitState.getProp(key));
+    }
+
+    String fileSplitBytesStr = workUnitState.getProp(HadoopFileInputSource.FILE_SPLIT_BYTES_STRING_KEY);
+    FileSplit fileSplit = deserializeFileSplit(fileSplitBytesStr);
+    FileInputFormat<K, V> fileInputFormat = getFileInputFormat(workUnitState, jobConf);
+    RecordReader<K, V> recordReader = fileInputFormat.getRecordReader(fileSplit, jobConf, Reporter.NULL);
+    boolean readKeys = workUnitState.getPropAsBoolean(
+        HadoopFileInputSource.FILE_INPUT_READ_KEYS_KEY, HadoopFileInputSource.DEFAULT_FILE_INPUT_READ_KEYS);
+    return getExtractor(workUnitState, recordReader, fileSplit, readKeys);
+  }
+
+  @Override
+  public void shutdown(SourceState state) {
+
+  }
+
+  /**
+   * Get a {@link FileInputFormat} instance used to get {@link FileSplit}s and a {@link RecordReader}
+   * for every {@link FileSplit}.
+   *
+   * <p>
+   *   This default implementation simply creates a new instance of a {@link FileInputFormat} class
+   *   specified using the configuration property {@link HadoopFileInputSource#FILE_INPUT_FORMAT_CLASS_KEY}.
+   * </p>
+   *
+   * @param state a {@link State} object carrying configuration properties
+   * @param jobConf a Hadoop {@link JobConf} object carrying Hadoop configurations
+   * @return a {@link FileInputFormat} instance
+   */
+  @SuppressWarnings("unchecked")
+  protected FileInputFormat<K, V> getFileInputFormat(State state, JobConf jobConf) {
+    Preconditions.checkArgument(state.contains(HadoopFileInputSource.FILE_INPUT_FORMAT_CLASS_KEY));
+    try {
+      return (FileInputFormat<K, V>) ReflectionUtils.newInstance(
+          Class.forName(state.getProp(HadoopFileInputSource.FILE_INPUT_FORMAT_CLASS_KEY)), new Configuration());
+    } catch (ClassNotFoundException cnfe) {
+      throw new RuntimeException(cnfe);
+    }
+  }
+
+  /**
+   * Get a {@link OldApiHadoopFileInputExtractor} instance.
+   *
+   * @param workUnitState a {@link WorkUnitState} object carrying Gobblin configuration properties
+   * @param recordReader a Hadoop {@link RecordReader} object used to read input records
+   * @param fileSplit the {@link FileSplit} to read input records from
+   * @param readKeys whether the {@link OldApiHadoopFileInputExtractor} should read keys of type {@link #<K>};
+   *                 by default values of type {@link #>V>} are read.
+   * @return a {@link OldApiHadoopFileInputExtractor} instance
+   */
+  protected abstract OldApiHadoopFileInputExtractor<S, D, K, V> getExtractor(WorkUnitState workUnitState,
+      RecordReader<K, V> recordReader, FileSplit fileSplit, boolean readKeys);
+
+  private String serializeFileSplit(FileSplit fileSplit) throws IOException {
```

Comment:
Could you just move this method and `deserializeFileSplit` into a utility class, and have the parameter be of type `Writable`. These methods should work for any custom `Writable` class that needs to be serialized into a string.